lib/spack/llnl/util/filesystem.py

```diff
@@ -1446,26 +1465,33 @@ def visit_directory_tree(root, visitor, rel_path="", depth=0):
                 # will ensure that if so, that it is relative
                 # to the CWD and therefore
                 # makes sense
-                isdir = os.path.isdir(link_target)
+                # if the target of a symlink is another malformed directory symlink
+                # is_dir will throw again. Instead, check to see if the file is a symlink first
+                # if it's a symlink, even if a symlink to a dir, it's still pointing
+                # to a symlink and is therefor itself, not a dir
+                # if it's pointing to a valid symlink to a dir, previous behavior would
+                # return false, so if link_target is a symlink, we should always report false
+                # this keeps behavior consistent with previous os.path approach
+                isdir = False if link_target.is_symlink() else link_target.is_dir()
```

Comment:
Thank you for this thorough explanation of the logic! We'll probably want to reword this a little bit before merging, but let's leave that for later.

Reply:
Glad you found it useful, it is a _little_ wordy/meandering, agree on the refactor